plugins/modules/vmware_guest_network.py

```diff
@@ -250,316 +337,442 @@ def __init__(self, module):
+            nic_spec = vim.vm.device.VirtualDeviceSpec(
+                device=device_obj()
+            )
+            if mac_address:
+                nic_spec.device.addressType = 'manual'
+                nic_spec.device.macAddress = mac_address
+
+            if label:
+                nic_spec.device.deviceInfo = vim.Description(
+                    label=label
+                )
         else:
-            nic.device.addressType = 'generated'
-        if 'directpath_io' in device_info:
-            if isinstance(nic.device, vim.vm.device.VirtualVmxnet3):
-                nic.device.uptCompatibilityEnabled = device_info['directpath_io']
-            else:
-                self.module.fail_json(msg='UPT is only compatible for Vmxnet3 adapter.'
-                                      + ' Clients can set this property enabled or disabled if ethernet virtual device is Vmxnet3.')
-
-        return nic
-
-    def get_network_info(self, vm_obj):
-        network_info = dict()
-        if vm_obj is None:
-            return network_info
-
-        nic_index = 0
-        for nic in vm_obj.config.hardware.device:
-            nic_type = None
-            directpath_io = 'N/A'
-            if isinstance(nic, vim.vm.device.VirtualPCNet32):
-                nic_type = 'PCNet32'
-            elif isinstance(nic, vim.vm.device.VirtualVmxnet2):
-                nic_type = 'VMXNET2'
-            elif isinstance(nic, vim.vm.device.VirtualVmxnet3):
-                nic_type = 'VMXNET3'
-                directpath_io = nic.uptCompatibilityEnabled
-            elif isinstance(nic, vim.vm.device.VirtualE1000):
-                nic_type = 'E1000'
-            elif isinstance(nic, vim.vm.device.VirtualE1000e):
-                nic_type = 'E1000E'
-            elif isinstance(nic, vim.vm.device.VirtualSriovEthernetCard):
-                nic_type = 'SriovEthernetCard'
-            if nic_type is not None:
-                network_info[nic_index] = dict(
-                    device_type=nic_type,
-                    label=nic.deviceInfo.label,
-                    name=nic.deviceInfo.summary,
-                    mac_addr=nic.macAddress,
-                    unit_number=nic.unitNumber,
-                    wake_onlan=nic.wakeOnLanEnabled,
-                    allow_guest_ctl=nic.connectable.allowGuestControl,
-                    connected=nic.connectable.connected,
-                    start_connected=nic.connectable.startConnected,
-                    directpath_io=directpath_io
+            nic_spec = vim.vm.device.VirtualDeviceSpec(
+                operation=vim.vm.device.VirtualDeviceSpec.Operation.edit,
+                device=nic_obj
+            )
+            if label and label != nic_obj.deviceInfo.label:
+                nic_spec.device.deviceInfo = vim.Description(
+                    label=label
                 )
-                nic_index += 1
-
-        return network_info
-
-    def sanitize_network_params(self):
```

Comment:
there are a bunch of params check in this method. Please make sure you are adding adequate checks and fail with appropriate error messages.
For example, 

- vm not found
- if `directpath_io` is set for other than `Vmxnet3 adapter`
- check for this condition
```
cluster:
    description:
      - Name of cluster where vm is run
      - required if esxi_hostname isn't set
```

Reply:
For some reason I assumed `vm not found`was handled in the ansible pyvmomi helper, fixed. Regarding directpath_io I had some reasoning that it wasn't needed to fail, just silently ignore it, but I can't remember why and it seems like bad reasoning. I've added cluster and esxi_hostname to required_one_of

Reply:
`cluster:
    description:
      - Name of cluster where vm is run
      - required if esxi_hostname isn't set
`
Actually, this is wrong, cluster isn't needed, it may be so that either parameters hostname and esxi_hostname is needed, this may be covered in the parent (vmware) documentation?